### PR TITLE
[tf2] cherry pick: Move some packages to tf2

### PR DESCRIPTION
**Pull request by moriarty**
This cherry-picks commit 144cae29
https://github.com/ros-planning/navigation/pull/531 from @vrabaud which targeted Kinetic.

I fixed the conflicts.

Conflicts:
	base_local_planner/package.xml
	base_local_planner/src/trajectory_planner_ros.cpp
	carrot_planner/package.xml
	global_planner/src/orientation_filter.cpp

 Author:    Vincent Rabaud
 Date:      Mon Oct 24 22:08:55 2016 +0200

### CODE REVIEW

**Comment by moriarty** [comment]
@mikeferguson & @DLu, cherry-picking this looked like low hanging fruit for:
https://github.com/ros-planning/navigation/milestone/4

**Comment by moriarty** [comment]
It looks like #458 could similarly be cherry-picked or rebased into melodic. It was just a bigger conflict list so I didn't do it last night.

https://github.com/ros-planning/navigation/pull/561#issuecomment-319017774

**Comment by mikeferguson** [comment]
How have you tested this? I have yet to have a TF2 PR for the costmaps/planners that actually worked on a robot (or in sim).

**Comment by moriarty** [comment]
@mikeferguson I did some basic system level testing of this indirectly by launching: https://github.com/fetchrobotics/fetch_ros/blob/melodic-devel/fetch_navigation/launch/fetch_nav.launch and giving it some nav goals... It seemed okay.

I didn't test the carrot_planner ;)

But I didn't do any side by side testing to compare the differences with and without this PR.

I'm currently cherry-picking the #458 changes and can do some more testing when done that.

**Comment by moriarty** [comment]
FYI: @jcmayoral, @oscar-lima, if you have time to give this some more testing on other hardware.

File: base_local_planner/CMakeLists.txt [hunk 1]
```diff
@@ -18,6 +18,7 @@ find_package(catkin REQUIRED
             rospy
             std_msgs
             tf
```
Comment (mikeferguson) [inline base_local_planner/CMakeLists.txt:20]: shouldn't we be getting rid of this if we are moving to tf2?

**Review by mikeferguson — APPROVED** [review]
Ok -- so this just does the planners -- I can confirm that this is working for me (on simulated robot).

If I recall correctly, costmap2d is where things got wonky -- but somehow I missed that we can do these planners separately from costmap2d.

I'm gonna give this some additional comment time (til monday) before I merge.

**Comment by mikeferguson** [comment]
@moriarty do you plan on having time to work on the costmap2d -> tf2 part? If not, perhaps we'll move forward with a Melodic release without that conversion having been done, and push that to N-turtle.

**Comment by moriarty** [comment]
@mikeferguson yes I have been testing this and also was trying to get @vrabaud's other branch from #458 which does the rest of the tf2 changes... I was busy on something else during the week and that's why I tagged two other ppl who I thought might be able to try it out in their labs.

I was hoping to finish this weekend, sorry for the delay, I got side tracked with two other issues... occasionally https://github.com/ros-perception/slam_karto crashes on loop closure, and the PS3 Joy isn't great at the moment.

Here is the line you mentioned above:
https://github.com/ros-planning/navigation/pull/458/files#diff-ed5ea4540bc099d8f4287c266e2ffafcL20

**Comment by mikeferguson** [comment]
Just a note -- the crash on loop closure may not be all that new if it is intermittent...

**Comment by oscar-lima** [comment]
I confirm it works for us in real robot and simulation, but under kinetic (we do not have melodic) #531

**Comment by jcmayoral** [comment]
Sorry for the delay. I tested on simulation and works well (under kinetic).

**Comment by moriarty** [comment]
screenshot from 2018-07-14 15-40-30
screenshot from 2018-07-14 15-42-59
screenshot from 2018-07-14 15-43-31
screenshot from 2018-07-14 15-43-59
screenshot from 2018-07-14 15-44-41

**Comment by moriarty** [comment]
I checked this branch out and gave it some more testing on the hardware in a clean ros workspace. 
Everything seemed ok- the costmap visualization depends on zooming around and pan/tilting in RViz. See screenshots

**Comment by moriarty** [comment]
Ignore my above weird rviz visualization comment
screenshot from 2018-07-14 15-50-34
screenshot from 2018-07-14 15-52-09

Setting the Alpha on all maps to 0.7 and the global map "draw behind" fixed any weird visualization errors so I don't thing that is the "wonky" costmap thing you were referring to.

**Comment by mikeferguson** [comment]
My comment about cost maps was that the costmap2d package is still using TF, having not yet been ported to TF2. I've previously worked on this a bit, but had issues trying to port it to TF2.

Thank you to all for the feedback -- I'd say this is probably the most tested PR we've had recently.

**Comment by moriarty** [comment]
@mikeferguson, :+1: yes cherry-picking the other change with all the changes for melodic at the same time in my workspace led to some pain... I'll give it another shot now that this one is merged.

**Comment by moriarty** [comment]
thanks @jcmayoral && @oscar-lima for testing ;) but you should upgrade to 18.04 & Melodic :laughing:

**Comment by mikeferguson** [comment]
Not sure why -- but CI timed out after merging this. http://build.ros.org/job/Mdev__navigation__ubuntu_bionic_amd64/14/

I've retriggered to see if is a Jenkins thing...

**Comment by moriarty** [comment]
@mikeferguson I will look into it

**Comment by moriarty** [comment]
My fault... sorry I've gotten use to tests being run on the branches of pull requests...

gtest-all.cc:(.text._ZNK7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEE16GetOrCreateValueEv[_ZNK7testing8internal11ThreadLocalIPNS_31TestPartResultReporterInterfaceEE16GetOrCreateValueEv]+0x88): undefined reference to `pthread_setspecific'
```
[build] Summary: 9 of 15 packages succeeded.                                                                                                                                                               
[build]   Ignored:   1 packages were skipped or are blacklisted.                                                                                                                                           
[build]   Warnings:  1 packages succeeded with warnings.                                                                                                                                                   
[build]   Abandoned: 5 packages were abandoned.                                                                                                                                                            
[build]   Failed:    1 packages failed.                                                                                                                                                                    
[build] Runtime: 4 minutes and 25.7 seconds total. 
```
With this change, 14/15 tests have passed and AMCL is still running
```
diff --git a/base_local_planner/CMakeLists.txt b/base_local_planner/CMakeLists.txt
index abad7e34..0b1b02fa 100644
--- a/base_local_planner/CMakeLists.txt
+++ b/base_local_planner/CMakeLists.txt
@@ -141,6 +141,8 @@ install(DIRECTORY include/${PROJECT_NAME}/
 
 if(CATKIN_ENABLE_TESTING)
   find_package(rostest)
+  SET(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -pthread")
+
   catkin_add_gtest(base_local_planner_utest
     test/gtest_main.cpp
     test/utest.cpp

```

File: carrot_planner/src/carrot_planner.cpp [hunk 2]
```diff
@@ -155,7 +149,8 @@ namespace carrot_planner {
 
     plan.push_back(start);
     geometry_msgs::PoseStamped new_goal = goal;
-    tf::Quaternion goal_quat = tf::createQuaternionFromYaw(target_yaw);
+    tf2::Quaternion goal_quat;
+    goal_quat.setEuler(target_yaw, 0, 0);
```
Comment (mikeferguson) [inline carrot_planner/src/carrot_planner.cpp:153]: This is one potential issue -- it looks like internally createQuaternionFromYaw() called setRPY(). In TF2 -- setEuler() is different from SetRPY() or SetEulerZYX() which internally calls SetRPY()

File: global_planner/src/orientation_filter.cpp [hunk 3]
```diff
@@ -35,19 +35,17 @@
...
 void set_angle(geometry_msgs::PoseStamped* pose, double angle)
 {
-    pose->pose.orientation = tf::createQuaternionMsgFromYaw(angle); 
-}
-
-double getYaw(geometry_msgs::PoseStamped pose)
-{
-    return tf::getYaw(pose.pose.orientation);
+  tf2::Quaternion q;
+  q.setEuler( angle, 0, 0 );
```
Comment (mikeferguson) [inline global_planner/src/orientation_filter.cpp:47]: again, I think this should be SetRPY possibly?

File: base_local_planner/src/trajectory_planner_ros.cpp [hunk 4]
```diff
@@ -511,14 +511,15 @@ namespace base_local_planner {
...
-              global_frame_);
       geometry_msgs::PoseStamped pose;
-      tf::poseStampedTFToMsg(p, pose);
+      pose.header.frame_id = global_frame_;
+      pose.header.stamp = ros::Time::now();
+      pose.pose.position.x = p_x;
+      pose.pose.position.y = p_y;
+      pose.pose.position.z = 0.0;
+      tf2::Quaternion q;
+      q.setEuler(p_th, 0, 0);
```
Comment (mikeferguson) [inline base_local_planner/src/trajectory_planner_ros.cpp:521]: same note -- setRPY might be the correct thing here?

File: global_planner/src/orientation_filter.cpp (hunk 3 above)
Comment (reinzor) [reply global_planner/src/orientation_filter.cpp:47]: I can verify that changing `q.setEuler( angle, 0, 0 );` to `q.setRPY(0, 0, angle);` works. This fixes https://github.com/ros-planning/navigation/issues/760

**Pull request merged**